crates/rustc_codegen_spirv/src/builder/load_store.rs

```diff
@@ -0,0 +1,506 @@
+                    field_types.iter().zip(field_offsets.iter()).enumerate()
+                {
+                    let load_res = self.extract_value(val, element_idx as u64);
+
+                    if offset.bytes() as u32 % 4 != 0 {
+                        let adt_name = self.type_cache.lookup_name(val.ty);
+                        let field_name = if let Some(field_names) = field_names {
+                            &field_names[element_idx]
+                        } else {
+                            "<unknown>"
+                        };
+
+                        self.fatal(&format!(
+                            "Trying to store to unaligned field: `{}::{}`. Field must be aligned to multiple of 4 bytes, but has offset {}",
+                            adt_name,
+                            field_name,
+                            offset.bytes() as u32));
+                    }
+
+                    let offset = offset.bytes() as u32 / 4;
+
+                    self.recurse_adt_for_stores(
+                        uint_ty,
+                        load_res,
+                        base_offset + offset,
+                        uint_values_and_offsets,
+                    );
+                }
+            }
+            SpirvType::Vector { count, element } => {
+                for offset in 0..count {
+                    let load_res = self.extract_value(val, offset as u64);
+
+                    self.recurse_adt_for_stores(
+                        uint_ty,
+                        load_res,
+                        base_offset + offset,
+                        uint_values_and_offsets,
+                    );
+                }
+            }
+            SpirvType::Float(bits) => {
+                let unsigned_ty = SpirvType::Integer(bits, false).def(rustc_span::DUMMY_SP, self);
+                let val_def = val.def(self);
+
+                let bitcast_res = self
+                    .emit()
+                    .bitcast(unsigned_ty, None, val_def)
+                    .unwrap()
+                    .with_type(unsigned_ty);
+
+                self.store_as_u32(
+                    bits,
+                    false,
+                    uint_ty,
+                    bitcast_res,
+                    base_offset,
+                    uint_values_and_offsets,
+                );
+            }
+            SpirvType::Integer(bits, signed) => {
+                self.store_as_u32(
+                    bits,
+                    signed,
+                    uint_ty,
+                    val,
+                    base_offset,
+                    uint_values_and_offsets,
+                );
+            }
+            _ => {
+                let mut err = self.tcx.sess.struct_err(
+                    "Type unsupported for `codegen_internal_buffer_store` return / args:",
+                );
+                err.note(&format!("type `{:?}`", val));
+                err.emit();
+            }
```

Comment:
We should support more types here - e.g. sized arrays. Listing every type and being explicit about what we don't support is probably best for maintainability when adding new types.

Reply:
Added support for sized arrays, and list all of the types individually to provide better error messages.